From 4a19662306793c25c93b02a4420bd175f707f41c Mon Sep 17 00:00:00 2001
From: Chen Jiahao <chenjiahao16@huawei.com>
Date: Sat, 12 Nov 2022 23:50:03 +0800
Subject: [PATCH 0035/1398] [Huawei] arm64: Introduce
 CONFIG_DUMP_UNDEFINE_INSTR

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187967

-------------------------------

Introduce CONFIG_DUMP_UNDEFINE_INSTR to isolate feature
which providing the ability to dump regs and instruction
when undefined instruction exception occurs.

Signed-off-by: Chen Jiahao <chenjiahao16@huawei.com>
---
 arch/arm64/Kconfig                     | 9 +++++++++
 arch/arm64/configs/openeuler_defconfig | 1 +
 arch/arm64/kernel/traps.c              | 4 ++++
 3 files changed, 14 insertions(+)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 72743b5ed3e3..ff15ea9a27a4 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -1268,6 +1268,15 @@ config CRASH_DUMP
 
 	  For more details see Documentation/admin-guide/kdump/kdump.rst
 
+config DUMP_UNDEFINED_INSTR
+	bool "dump undefined instruction before BUG"
+	default n
+	help
+	  In case of debugging need when undefined instructin exception
+	  occurs, we should dump regs and the undefined instruction
+	  before BUG in kernel mode. Which is usually used to compare
+	  with objdump instructions.
+
 config ARM64_CPU_PARK
 	bool "Support CPU PARK on kexec"
 	depends on SMP
diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 2cd41b103121..be3c51633c4b 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -434,6 +434,7 @@ CONFIG_KEXEC_FILE=y
 CONFIG_KEXEC_SIG=y
 CONFIG_KEXEC_IMAGE_VERIFY_SIG=y
 CONFIG_CRASH_DUMP=y
+CONFIG_DUMP_UNDEFINED_INSTR=y
 CONFIG_ARM64_CPU_PARK=y
 # CONFIG_XEN is not set
 CONFIG_FORCE_MAX_ZONEORDER=11
diff --git a/arch/arm64/kernel/traps.c b/arch/arm64/kernel/traps.c
index f8ec533eb590..6a9f234587c4 100644
--- a/arch/arm64/kernel/traps.c
+++ b/arch/arm64/kernel/traps.c
@@ -405,8 +405,12 @@ void do_undefinstr(struct pt_regs *regs)
 	if (call_undef_hook(regs) == 0)
 		return;
 
+#ifdef CONFIG_DUMP_UNDEFINED_INSTR
 	if (!user_mode(regs))
 		die("Oops - undefined instruction", regs, 0);
+#else
+	BUG_ON(!user_mode(regs));
+#endif
 
 	force_signal_inject(SIGILL, ILL_ILLOPC, regs->pc, 0);
 }
-- 
2.17.1

